libzeth/core/utils.cpp

```diff
@@ -83,28 +112,21 @@ std::string bytes_to_hex(const std::string &s)
     return out;
 }
 
-int hex_to_bytes(char *source_str, uint8_t *dest_buffer)
+std::string bytes_to_hex_reversed(const void *bytes, size_t num_bytes)
 {
-    char *line = source_str;
-    char *data = line;
-    int offset;
-    int read_byte;
-    int data_len = 0;
-
-    while (sscanf(data, "%02x%n", &read_byte, &offset) == 1) {
-        dest_buffer[data_len++] = read_byte;
-        data += offset;
-    }
-    return data_len;
-}
+    std::string out;
+    out.reserve(num_bytes * 2);
 
-void erase_substring(std::string &string, const std::string &substring)
-{
-    size_t position = string.find(substring);
+    const uint8_t *const src_bytes_end = (const uint8_t *)bytes;
+    const uint8_t *src_bytes = src_bytes_end + num_bytes;
+    do {
+        --src_bytes;
+        const uint8_t byte = *src_bytes;
+        out.push_back(nibble_hex(byte >> 4));
+        out.push_back(nibble_hex(byte & 0x0f));
+    } while (src_bytes > src_bytes_end);
```

Comment:
You mean the do-while?  It can be a little less awkward to do decrementing loops like this. We don't have to setup the pointer as `dest_bytes + bytes - 1` or use the `>=` condition which is less safe with indices and pointers. (admittedly, with pointers it would only be an issue if `src_bytes_end == 0` which should never happen on any regular platform, but in general this form seemed more sensible when decrementing).

(Let me know if you are referring to something else here)

Reply:
> You mean the do-while?

Yes

Reply:
As pointed out by @dtebbs `i` is `size_t` so the condition above is always true for unsigned ints. We'd need to switch `i` to be signed for the loop to end